File: cpp/libjoynr/joynr-messaging/RoutingTable.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

#include "joynr/RoutingTable.h"
#include "joynr/system/RoutingTypes/MqttAddress.h"

namespace joynr
{

RoutingTable::RoutingTable(const std::string& gcdParticipantId)
        : _multiIndexContainer(), _gcdParticipantId(gcdParticipantId)
{
}

RoutingTable::~RoutingTable()
{
    JOYNR_LOG_TRACE(logger(), "destructor: number of entries = {}", _multiIndexContainer.size());
}

boost::optional<routingtable::RoutingEntry> RoutingTable::lookupRoutingEntryByParticipantId(
        const std::string& participantId) const
{
    auto& index = boost::multi_index::get<routingtable::tags::ParticipantId>(_multiIndexContainer);
    auto found = index.find(participantId);
    if (found == index.end()) {
        return boost::none;
    }
    return *found;
}

boost::optional<routingtable::RoutingEntry> RoutingTable::lookupRoutingEntryByParticipantIdAndGbid(
        const std::string& participantId,
        const std::string& gbid) const
{
    auto found = lookupRoutingEntryByParticipantId(participantId);
    if (found && (participantId == this->_gcdParticipantId)) {
        auto address = found->address;
        if (auto mqttAddress =
                    dynamic_cast<const joynr::system::RoutingTypes::MqttAddress*>(address.get())) {
            const auto newMqttAddress = std::make_shared<joynr::system::RoutingTypes::MqttAddress>(
                    gbid, mqttAddress->getTopic());
            return routingtable::RoutingEntry(participantId,
                                              newMqttAddress,
                                              found->isGloballyVisible,
                                              found->_expiryDateMs,
                                              found->_isSticky);
        }
    }
    return found;
}

std::unordered_set<std::string> RoutingTable::lookupParticipantIdsByAddress(
        std::shared_ptr<const joynr::system::RoutingTypes::Address> searchValue) const
{
    std::unordered_set<std::string> result;
    const auto& addressIndex =
            boost::multi_index::get<routingtable::tags::Address>(_multiIndexContainer);
    auto found = addressIndex.equal_range(searchValue);
    for (auto it = found.first; it != found.second; ++it) {
        result.insert(it->participantId);
    }
    return result;
}

bool RoutingTable::containsParticipantId(const std::string& participantId) const
{
    auto& index = boost::multi_index::get<routingtable::tags::ParticipantId>(_multiIndexContainer);
    auto found = index.find(participantId);
    return found != index.end();
}

void RoutingTable::add(const std::string& participantId,
                       bool isGloballyVisible,
                       std::shared_ptr<const joynr::system::RoutingTypes::Address> address,
                       std::int64_t expiryDateMs,
                       bool isSticky)
{
    routingtable::RoutingEntry routingEntry(participantId,
                                            std::move(address),
                                            std::move(isGloballyVisible),
                                            std::move(expiryDateMs),
                                            std::move(isSticky));
    auto result = _multiIndexContainer.insert(routingEntry);
    if (!result.second) {
        _multiIndexContainer.replace(result.first, routingEntry);
        JOYNR_LOG_INFO(logger(),
                       "Replaced routing entry: new: {}, old: {}, #entries: {}",
                       routingEntry.toString(),
                       result.first->toString(),
                       _multiIndexContainer.size());
    } else {
        JOYNR_LOG_INFO(logger(),
                       "Added routing entry: {}, #entries: {}",
                       routingEntry.toString(),
                       _multiIndexContainer.size());
    }
}

void RoutingTable::remove(const std::string& participantId)
{
    const auto routingEntry = lookupRoutingEntryByParticipantId(participantId);
    if (routingEntry && routingEntry->_isSticky) {
        JOYNR_LOG_WARN(logger(),
                       "Cannot remove sticky routing entry (participantId={}, address={}, "
                       "isGloballyVisible={}, expiryDateMs={}) from routing table",
                       participantId,
                       routingEntry->address->toString(),
                       routingEntry->isGloballyVisible,
                       routingEntry->_expiryDateMs);
        return;
    }
    JOYNR_LOG_INFO(logger(),
                   "Removing routing entry for participantId: {}, #entries before removal: {}",
                   participantId,
                   _multiIndexContainer.size());
    _multiIndexContainer.erase(participantId);
}

void RoutingTable::purge()
{
    bool expiredEntriesFound = false;
    auto& index = boost::multi_index::get<routingtable::tags::ExpiryDate>(_multiIndexContainer);
    auto now = std::chrono::duration_cast<std::chrono::milliseconds>(
                       std::chrono::system_clock::now().time_since_epoch()).count();
    auto last = index.upper_bound(now);
    std::vector<std::string> expiredParticipantIds;
    for (auto routingEntryIterator = index.lower_bound(0); routingEntryIterator != last;
         ++routingEntryIterator) {
        if (!routingEntryIterator->_isSticky) {
            expiredParticipantIds.push_back(routingEntryIterator->participantId);
            expiredEntriesFound = true;
        }
    }
    if (expiredEntriesFound) {
        JOYNR_LOG_INFO(logger(), "Purging expired routing entries");
    }
    for (auto& participantId : expiredParticipantIds) {
        remove(participantId);
    }
}

bool RoutingTable::AddressEqual::operator()(
        std::shared_ptr<const joynr::system::RoutingTypes::Address> lhs,
        std::shared_ptr<const joynr::system::RoutingTypes::Address> rhs) const
{
    bool result = ((*lhs) == (*rhs));
    return result;
}

size_t RoutingTable::AddressHash::operator()(
        std::shared_ptr<const joynr::system::RoutingTypes::Address> address) const
{
    return address->hashCode();
}

} // namespace joynr

File: java/messaging/messaging-common/src/main/java/io/joynr/messaging/routing/RoutingTable.java
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.messaging.routing;

import joynr.system.RoutingTypes.Address;

public interface RoutingTable {
    /**
     * @param participantId participantId for which an Address will be returned from the
     * routing table
     * @return Address the stored address for the given participantId (unmodified, see {@link #get(String, String)}
     */
    Address get(String participantId);

    /**
     * @param participantId participantId for which an Address will be returned from the
     * routing table
     * @param gbid name of the backend (evaluated for gcdParticipantId only):
     * selects the backend specific address of GCD (if participantId is the participantId of the GCD)
     * @return Address the stored address for the given participantId<br>
     * - in case of gcdParticipantId, the gbid is evaluated to adapt the stored address for the selected
     *   backend before it is returned (this allows communication with GCD instances in different backends<br>
     *   even though there is only one address per participantId<br>
     *   return null if the gbid is unknown (not part of the configured list of
     *   GBIDs {@link io.joynr.messaging.ConfigurableMessagingSettings#PROPERTY_GBIDS}<br>
     * - return unmodified address otherwise
     */
    Address get(String participantId, String gbid);

    /**
     * Sets the participantId of the Global Capabilities Directory (GCD) before adding it to the routing table
     * @param gcdParticipantId the participantId of the GCD
     */
    public void setGcdParticipantId(final String gcdParticipantId);

    /**
     * Adds a new routing entry. If a routing entry for the provided participantId already exists, only the expiryDate
     * and the sticky-flag are updated unless the update is allowed (See RoutingTableAddressValidator).
     *
     * @param participantId participant id for which a routing entry shall be created
     * @param address Address which shall be associated with the participant id
     * @param isGloballyVisible States whether the endpoint is globally visible or not
     * @param expiryDateMs Expiry date of the routing entry in milliseconds
     * @param isSticky If set to true, the routing entry never expires and cannot be replaced
     */
    void put(String participantId, Address address, boolean isGloballyVisible, long expiryDateMs, boolean isSticky);

    /**
     * Overload of put method with isSticky set to false.
     *
     * @param participantId participant id for which a routing entry shall be created
     * @param address Address which shall be associated with the participant id
     * @param isGloballyVisible States whether the endpoint is globally visible or not
     * @param expiryDateMs Expiry date of the routing entry in milliseconds
     */
    void put(String participantId, Address address, boolean isGloballyVisible, long expiryDateMs);

    boolean containsKey(String participantId);

    /**
     * Query the routing table for the status of isGloballyVisible parameter
     * @param participantId participantId for which the visibility shall be looked up
     * @return true if participantId is globally visible,
     *         false if participantId is not globally visible
     * @throws io.joynr.exceptions.JoynrRuntimeException if no entry exists for the given participantId
     */
    boolean getIsGloballyVisible(String participantId);

    /**
     * Query the expiry date of a routing entry for a participant id.
     * @param participantId participantId for which the expiryDate shall be looked up
     * @return The routing entry's expiry date in ms.
     */
    long getExpiryDateMs(String participantId);

    /**
     * Query the sticky-flag of a routing entry for a participant id.
     * @param participantId participantId for which the sticky-flag shall be looked up
     * @return The routing entry's sticky state.
     */
    boolean getIsSticky(String participantId);

    void remove(String participantId);

    /**
     * Apply the specified operation to all addresses currently held in the routing table.
     *
     * @param addressOperation
     *            the address operation to perform.
     */
    void apply(AddressOperation addressOperation);

    /**
     * Purge all expired routing entries from the table
     */
    void purge();
}
